tx-pool/src/process.rs

```diff
@@ -1032,43 +1032,40 @@ fn _update_tx_pool_for_reorg(
         let mut gaps = Vec::new();
 
         for entry in tx_pool.pool_map.entries.get_by_status(&Status::Gap) {
-            let e = &entry.inner;
-            let short_id = e.proposal_short_id();
+            let short_id = entry.inner.proposal_short_id();
             if snapshot.proposals().contains_proposed(&short_id) {
-                proposals.push(e.clone());
+                proposals.push(short_id);
             }
         }
 
         for entry in tx_pool.pool_map.entries.get_by_status(&Status::Pending) {
-            let e = &entry.inner;
-            let short_id = e.proposal_short_id();
+            let short_id = entry.inner.proposal_short_id();
             if snapshot.proposals().contains_proposed(&short_id) {
-                proposals.push(e.clone());
+                proposals.push(short_id);
             } else if snapshot.proposals().contains_gap(&short_id) {
-                gaps.push(e.clone());
+                gaps.push(short_id);
             }
         }
 
-        for entry in proposals {
-            debug!("begin to proposed: {:x}", entry.transaction().hash());
-            let cached = CacheEntry::completed(entry.cycles, entry.fee);
-            if let Err(e) =
-                tx_pool.proposed_rtx(cached, entry.size, entry.timestamp, Arc::clone(&entry.rtx))
-            {
-                callbacks.call_reject(tx_pool, &entry, e.clone());
+        for id in proposals {
+            debug!("begin to proposed: {:x}", id);
+            let entry = tx_pool.get(&id).unwrap().clone();
+            if let Err(e) = tx_pool.proposed_rtx(&id) {
+                callbacks.call_reject(tx_pool, &entry, e);
             } else {
-                callbacks.call_proposed(tx_pool, &entry, false);
+                callbacks.call_proposed(tx_pool, &entry, false)
             }
         }
 
-        for entry in gaps {
-            debug!("begin to gap: {:x}", entry.transaction().hash());
-            let tx_hash = entry.transaction().hash();
-            let cached = CacheEntry::completed(entry.cycles, entry.fee);
-            if let Err(e) =
-                tx_pool.gap_rtx(cached, entry.size, entry.timestamp, Arc::clone(&entry.rtx))
-            {
-                debug!("Failed to add tx to gap {}, reason: {}", tx_hash, e);
+        for id in gaps {
+            debug!("begin to gap: {:x}", id);
+            if let Err(e) = tx_pool.gap_rtx(&id) {
+                let entry = tx_pool.get(&id).unwrap().clone();
```

Comment:
I'm concern about the `unwrap()` here, how about define `proposals` and `gaps` as below:
```rust
        let mut proposals: Vec<(ProposalShortId, Byte32)> = Vec::new();
        let mut gaps: Vec<(ProposalShortId, Byte32)> = Vec::new();
```
...
```rust
        for (id,entry) in proposals {
            debug!("begin to proposed: {:x}", id);
            if let Err(e) = tx_pool.proposed_rtx(&id) {
                callbacks.call_reject(tx_pool, &entry, e);
            } else {
                callbacks.call_proposed(tx_pool, &entry, false)
            }
        }

```

Reply:
I considered this here, the issue here is `tx_pool.proposed_rtx(&id)` is a `mut` reference here, then any later immutable borrow will be invalid.

All the `id` are got from `entries`, and if we re-query with `id` and get nothing, a `panic` is accepted since something weird and unexpected happened?

Reply:
I also want to avoid the `clone`, but haven't got a good solution right now.

Reply:
It's hard, how about this:
```rust
    if mine_mode {
        let mut proposals: Vec<(ProposalShortId, TxEntry)> = Vec::new();
        let mut gaps: Vec<(ProposalShortId, TxEntry)> = Vec::new();

        for entry in tx_pool.pool_map.entries.get_by_status(&Status::Gap) {
            let short_id = entry.inner.proposal_short_id();
            if snapshot.proposals().contains_proposed(&short_id) {
                proposals.push((short_id, entry.inner.to_owned()));
            }
        }

        for entry in tx_pool.pool_map.entries.get_by_status(&Status::Pending) {
            let short_id = entry.inner.proposal_short_id();
            if snapshot.proposals().contains_proposed(&short_id) {
                proposals.push((short_id, entry.inner.to_owned()));
            } else if snapshot.proposals().contains_gap(&short_id) {
                gaps.push((short_id, entry.inner.to_owned()));
            }
        }

        for (id, entry) in proposals {
            debug!("begin to proposed: {:x}", id);
            if let Err(e) = tx_pool.proposed_rtx(&id) {
                callbacks.call_reject(tx_pool, &entry, e);
            } else {
                callbacks.call_proposed(tx_pool, &entry, false)
            }
        }

        for (id, entry) in gaps {
            debug!("begin to gap: {:x}", id);
            if let Err(e) = tx_pool.gap_rtx(&id) {
                debug!(
                    "Failed to add tx to gap {}, reason: {}",
                    entry.transaction().hash(),
                    e
                );
                callbacks.call_reject(tx_pool, &entry, e.clone());
            }
        }
    }
```

Reply:
I think it's ok, at least we can avoid another query and the `unwrap`.